Mathlib/Algebra/Ring/Ext.lean

```diff
@@ -0,0 +1,477 @@
+/-
+Copyright (c) 2023 Raghuram Sundararajan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Raghuram Sundararajan
+-/
+import Mathlib.Algebra.Ring.Defs
+import Mathlib.Algebra.Group.Ext
+-- import Mathlib
+
+/-!
+# Extensionality lemmas for rings and similar structures
+
+In this file we prove extensionality lemmas for the ring-like structures defined in
+`Algebra/Ring/Defs`, ranging from `NonUnitalNonAssocSemiring` to `CommRing`. These extensionality
+lemmas take the form of asserting that two algebraic structures on a type are equal whenever the
+addition and multiplication defined by them are both the same.
+
+## Implementation details
+
+We follow `Algebra/Group/Ext` in using the term `(letI := i; HMul.hMul : R → R → R)` to refer to the
+multiplcation specified by a typeclass instance `i` on a type `R` (and similarly for addition). We
+abbreviate these using some local notations.
+
+Since `Algebra/Group/Ext` proved several injectivity lemmas, we do so as well — even if sometimes we
+don't need them to prove extensionality.
+
+## Tags
+semiring, ring, extensionality
+-/
+
+local macro:65 lhs:term:65 "+[" type:term ", " inst:term "]" rhs:term:66 : term =>
+  `(term| (letI := $inst; ($lhs : $type) + ($rhs : $type) : $type))
+local macro:70 lhs:term:70 "*[" type:term ", " inst:term "]" rhs:term:71 : term =>
+  `(term| (letI := $inst; ($lhs : $type) * ($rhs : $type) : $type))
+
+universe u
+
+variable {R : Type u}
+
+/-! ### Distrib -/
+namespace Distrib
+
+@[ext] theorem ext ⦃inst₁ inst₂ : Distrib R⦄
+    (h_add : ∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
```

Comment:
In the other ext lemmas, we write
```suggestion
    (h_mul : (letI := inst₁; HAdd.hAdd : R → R → R) = (letI := inst₂; HAdd.hAdd : R → R → R)) 
```
Can we do the same here?

Reply:
I thought that in practice equality of functions would mostly be proved by extensionality and so this would "shift" the need to use `funext` from uses to here, making the lemmas more usable.

For reference, I found two uses of the extensionality lemmas from `Mathlib.Algebra.Group.Ext`, in both of which the argument is proved using `funext`:
- `Group.ext` used in `Mathlib.Topology.Homotopy.HomotopyGroup`, around line 526
-  `AddCommGroup.ext` used in `Mathlib.CategoryTheory.Preadditive.Biproducts`, around line 657

Of course, I can also go ahead and change everything to "function equality" from "application equality".

Reply:
By the way, I noticed that `Group.ext` and `CommGroup.ext` in `Mathlib/Algebra/Group/Ext.lean` use `g.mul` rather than `(letI := g; HMul.hMul: G → G → G)`, which goes against what is written at the top of that file.

Should this be changed?

Reply:
What lemma does `@[ext]` generate for `Distrib`? Is it the same?

Reply:
The `@[ext]` in my code just registers the `ext` lemma I have proved for the `ext` tactic, as far as I know.

In case you mean what an `@[ext]` added to the definition of `Distrib` would do, it appears to be something more along the lines of `inst₁.toMul = inst₂.toMul → inst₁.toAdd = inst₂.toAdd → inst₁ = inst₂`.

Reply:
I agree with @urkud that we should follow the format of the `Distrib.ext` lemma which is generated by `@[ext] class Distrib`. In fact, [there is a library `note [partially-applied ext lemmas]` about this](https://leanprover-community.github.io/mathlib_docs/notes.html#partially-applied%20ext%20lemmas) (sadly never ported to Lean 4):

> When possible, `ext` lemmas are stated without a full set of arguments. As an example, for bundled
homs `f`, `g`, and `of`, `f.comp of = g.comp of → f = g` is a better `ext` lemma than
`(∀ x, f (of x) = g (of x)) → f = g`, as the former allows a second type-specific extensionality
lemmas to be applied to `f.comp of = g.comp of`.
If the domain of `of` is `ℕ` or `ℤ` and `of` is a `ring_hom`, such a lemma could then make the goal
`f (of 1) = g (of 1)`.
> 
> For bundled morphisms, there is a `ext` lemma that always applies of the form
`(∀ x, ⇑f x = ⇑g x) → f = g`. When adding type-specific `ext` lemmas like the one above, we want
these to be tried first. This happens automatically since the type-specific lemmas are inevitably
defined later.

So in this case we would use the unapplied `HAdd.add` and `HMul.mul`. (Note that `@[ext] class Distrib` is not quite right: it uses `Add.add` and `Mul.mul`, which means we don't get the nice notation.)

Reply:
In fact Eric has an in-depth explanation of why we shouldn't apply `ext` lemmas in his dissertation: https://eric-wieser.github.io/thesis/eric-wieser-thesis.pdf (chapter 5).

Reply:
That's a very good point for bundled morphisms, but I'm not sure that it applies to functions, where passing back from an "applied" goal `forall x y, f x y = g x y` to an "unapplied" goal `f = g` is a couple of `congr`s or so away for automation (unlike a bundled morphism, where one might have to reconstruct the bundled stuff if it has been simplified away), and as I mentioned above, it does seem like the "applied" goal is more often desired when used manually.

But in any case, I have changed everything to use "unapplied" ext subgoals in fbee3e7 (it was just a find-and-replace followed by patching proofs, but I don't *think* there are any errors in it now; still waiting for CI to finish).
(If this ever needs to be changed back, that and the three commits before it can be reverted.)